File: client/ts/src/table/actions.ts
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

import { actions } from "@/actions";
import {
  type Action,
  addCol,
  addRow,
  createReduceAll,
  type Handlers,
  removeCol,
  removeRow,
  rename,
  resizeCol,
  resizeRow,
  setCell,
} from "@/table/actions.gen";
import { type Cell } from "@/table/types.gen";

// MIN_CELL_DIM is the floor enforced on row and column sizes.
const MIN_CELL_DIM = 32;
// BASE_ROW_DIM and BASE_COL_DIM are the defaults used when an action
// bootstraps the opposing axis on an empty table.
const BASE_ROW_DIM = 36;
const BASE_COL_DIM = 72;

// deriveCellKey returns the key for the index-th replica of template. Both
// reducers run the same scheme so optimistic client state agrees with the
// server.
const deriveCellKey = (templateKey: string, index: number): string => {
  const suffix = index.toString(16).padStart(4, "0");
  if (templateKey.length < 36) return `${templateKey}-${suffix}`;
  return templateKey.slice(0, 32) + suffix;
};

// expandTemplate returns count replicas of template with keys derived via
// deriveCellKey.
const expandTemplate = (template: Cell, count: number): Cell[] =>
  Array.from({ length: count }, (_, i) => ({
    key: deriveCellKey(template.key, i),
    config: template.config,
  }));

const handlers: Handlers = {
  rename: (state, payload) => {
    const oldName = state.name;
    state.name = payload.name;
    return {
      inverse: [rename({ name: oldName })],
      targets: [state.key],
    };
  },

  addRow: (state, payload) => {
    const cells = payload.cellTemplate
      ? expandTemplate(
          payload.cellTemplate,
          state.columns.length === 0 && state.rows.length === 0
            ? 1
            : state.columns.length,
        )
      : payload.cells;
    if (state.columns.length === 0 && cells.length > 0)
      for (let i = 0; i < cells.length; i++) state.columns.push({ size: BASE_COL_DIM });
    const idx = Math.min(payload.index, state.rows.length);
    const keys = cells.map((c) => c.key);
    state.rows.splice(idx, 0, {
      size: Math.max(payload.size, MIN_CELL_DIM),
      cells: keys,
    });
    for (const c of cells) state.cells[c.key] = c.config;
    return {
      inverse: [removeRow({ index: idx })],
      targets: keys,
    };
  },

  removeRow: (state, payload) => {
    if (payload.index >= state.rows.length) return actions.NO_OP_RESULT;
    const removed = actions.snapshotDraft(state.rows[payload.index]);
    const cells: Cell[] = [];
    for (const k of removed.cells) {
      const c = state.cells[k];
      if (c != null) cells.push({ key: k, config: actions.snapshotDraft(c) });
    }
    state.rows.splice(payload.index, 1);
    for (const k of removed.cells) delete state.cells[k];
    return {
      inverse: [addRow({ index: payload.index, size: removed.size, cells })],
      targets: removed.cells,
    };
  },

  addCol: (state, payload) => {
    const cells = payload.cellTemplate
      ? expandTemplate(
          payload.cellTemplate,
          state.rows.length === 0 && state.columns.length === 0 ? 1 : state.rows.length,
        )
      : payload.cells;
    if (state.rows.length === 0 && cells.length > 0)
      for (let i = 0; i < cells.length; i++)
        state.rows.push({ size: BASE_ROW_DIM, cells: [] });
    const idx = Math.min(payload.index, state.columns.length);
    state.columns.splice(idx, 0, { size: Math.max(payload.size, MIN_CELL_DIM) });
    for (let i = 0; i < state.rows.length; i++) {
      if (i >= cells.length) break;
      const rowIdx = Math.min(idx, state.rows[i].cells.length);
      state.rows[i].cells.splice(rowIdx, 0, cells[i].key);
    }
    for (const c of cells) state.cells[c.key] = c.config;
    return {
      inverse: [removeCol({ index: idx })],
      targets: cells.map((c) => c.key),
    };
  },

  removeCol: (state, payload) => {
    if (payload.index >= state.columns.length) return actions.NO_OP_RESULT;
    const oldSize = state.columns[payload.index].size;
    const removedCells: Cell[] = [];
    state.columns.splice(payload.index, 1);
    // Rows missing a cell at this column are skipped, not padded. Go AddCol
    // short-circuits when len(cells) < len(rows), so the inverse repopulates
    // only the rows we captured.
    for (let i = 0; i < state.rows.length; i++) {
      if (payload.index >= state.rows[i].cells.length) continue;
      const k = state.rows[i].cells[payload.index];
      const c = state.cells[k];
      if (c != null) removedCells.push({ key: k, config: actions.snapshotDraft(c) });
      delete state.cells[k];
      state.rows[i].cells.splice(payload.index, 1);
    }
    return {
      inverse: [addCol({ index: payload.index, size: oldSize, cells: removedCells })],
      targets: removedCells.map((c) => c.key),
    };
  },

  resizeRow: (state, payload) => {
    if (payload.index >= state.rows.length) return actions.NO_OP_RESULT;
    const oldSize = state.rows[payload.index].size;
    state.rows[payload.index].size = Math.max(payload.size, MIN_CELL_DIM);
    return {
      inverse: [resizeRow({ index: payload.index, size: oldSize })],
      targets: [...state.rows[payload.index].cells],
    };
  },

  resizeCol: (state, payload) => {
    if (payload.index >= state.columns.length) return actions.NO_OP_RESULT;
    const oldSize = state.columns[payload.index].size;
    state.columns[payload.index].size = Math.max(payload.size, MIN_CELL_DIM);
    const targets = state.rows
      .map((r) => r.cells[payload.index])
      .filter((k): k is string => k != null);
    return {
      inverse: [resizeCol({ index: payload.index, size: oldSize })],
      targets,
    };
  },

  setCell: (state, payload) => {
    const existing = state.cells[payload.cell.key];
    if (existing == null) return actions.NO_OP_RESULT;
    const oldConfig = actions.snapshotDraft(existing);
    state.cells[payload.cell.key] = payload.cell.config;
    return {
      inverse: [setCell({ cell: { key: payload.cell.key, config: oldConfig } })],
      targets: [payload.cell.key],
    };
  },

  eraseCells: (state, payload) => {
    if (payload.cells.length === 0) return actions.NO_OP_RESULT;
    const selected = new Set(payload.cells);
    const rowPosOf = new Map<string, { row: number; col: number }>();
    for (let r = 0; r < state.rows.length; r++) {
      const cells = state.rows[r].cells;
      for (let c = 0; c < cells.length; c++) rowPosOf.set(cells[c], { row: r, col: c });
    }
    const rowTally = new Map<number, number>();
    const colTally = new Map<number, number>();
    for (const k of selected) {
      const pos = rowPosOf.get(k);
      if (pos == null) continue;
      rowTally.set(pos.row, (rowTally.get(pos.row) ?? 0) + 1);
      colTally.set(pos.col, (colTally.get(pos.col) ?? 0) + 1);
    }
    const fullRowIdx: number[] = [];
    for (const [rowIdx, tally] of rowTally)
      if (tally > 0 && tally === state.rows[rowIdx].cells.length)
        fullRowIdx.push(rowIdx);
    fullRowIdx.sort((a, b) => a - b);
    const fullColIdx: number[] = [];
    for (const [colIdx, tally] of colTally)
      if (tally === state.rows.length) fullColIdx.push(colIdx);
    fullColIdx.sort((a, b) => a - b);
    const inverse: Action[] = [];
    const targets: string[] = [];
    // Iterate highest-first so earlier removals don't shift later indices;
    // unshift the inverse so undo replays it in ascending order.
    for (let i = fullRowIdx.length - 1; i >= 0; i--) {
      const idx = fullRowIdx[i];
      const removed = actions.snapshotDraft(state.rows[idx]);
      const cells: Cell[] = [];
      for (const k of removed.cells) {
        const c = state.cells[k];
        if (c != null) cells.push({ key: k, config: actions.snapshotDraft(c) });
      }
      state.rows.splice(idx, 1);
      for (const k of removed.cells) delete state.cells[k];
      inverse.unshift(addRow({ index: idx, size: removed.size, cells }));
      targets.push(...removed.cells);
    }
    for (let i = fullColIdx.length - 1; i >= 0; i--) {
      const idx = fullColIdx[i];
      const oldSize = state.columns[idx].size;
      const removedCells: Cell[] = [];
      state.columns.splice(idx, 1);
      for (let r = 0; r < state.rows.length; r++) {
        if (idx >= state.rows[r].cells.length) continue;
        const k = state.rows[r].cells[idx];
        const c = state.cells[k];
        if (c != null) removedCells.push({ key: k, config: actions.snapshotDraft(c) });
        delete state.cells[k];
        state.rows[r].cells.splice(idx, 1);
      }
      inverse.unshift(addCol({ index: idx, size: oldSize, cells: removedCells }));
      targets.push(...removedCells.map((c) => c.key));
    }
    for (const k of selected) {
      const existing = state.cells[k];
      if (existing == null) continue;
      const oldConfig = actions.snapshotDraft(existing);
      state.cells[k] = payload.template;
      inverse.push(setCell({ cell: { key: k, config: oldConfig } }));
      targets.push(k);
    }
    return { inverse, targets };
  },
};

export const reduceAll = createReduceAll(handlers);

File: core/pkg/service/table/actions.go
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package table

import (
	"fmt"
	"slices"

	"github.com/synnaxlabs/x/set"
)

// minCellDim is the floor enforced on row and column sizes.
const minCellDim = 32

// baseRowDim and baseColDim are the defaults used when an action bootstraps
// the opposing axis on an empty table.
const (
	baseRowDim = 36
	baseColDim = 72
)

// deriveCellKey returns the key for the index-th replica of template. Both
// reducers run the same scheme so optimistic client state agrees with the
// server.
func deriveCellKey(templateKey string, index int) string {
	if len(templateKey) < 36 {
		return fmt.Sprintf("%s-%04x", templateKey, index)
	}
	return templateKey[:32] + fmt.Sprintf("%04x", index)
}

// expandTemplate returns count replicas of template with keys derived via
// deriveCellKey.
func expandTemplate(template Cell, count int) []Cell {
	cells := make([]Cell, count)
	for i := range cells {
		cells[i] = Cell{
			Key:    deriveCellKey(template.Key, i),
			Config: template.Config,
		}
	}
	return cells
}

// Handle replaces the table's name.
func (p RenamePayload) Handle(state Table) (Table, error) {
	state.Name = p.Name
	return state, nil
}

// Handle inserts a row at the given index. When CellTemplate is non-nil it
// replaces Cells with one replica per existing column (or one when the table
// is empty, which also seeds a default column). When both are absent against
// a column-less table, the new row is inserted with no cells. Sizes clamp
// up to minCellDim; indices clamp down to the end of the rows slice.
func (p AddRowPayload) Handle(state Table) (Table, error) {
	cells := p.Cells
	if p.CellTemplate != nil {
		n := len(state.Columns)
		if n == 0 && len(state.Rows) == 0 {
			n = 1
		}
		cells = expandTemplate(*p.CellTemplate, n)
	}
	if len(state.Columns) == 0 && len(cells) > 0 {
		state.Columns = slices.Repeat([]Column{{Size: baseColDim}}, len(cells))
	}
	keys := make([]string, len(cells))
	for i, c := range cells {
		keys[i] = c.Key
	}
	row := Row{Size: max(p.Size, minCellDim), Cells: keys}
	idx := min(int(p.Index), len(state.Rows))
	state.Rows = slices.Insert(state.Rows, idx, row)
	if state.Cells == nil {
		state.Cells = make(map[string]CellConfig, len(cells))
	}
	for _, c := range cells {
		state.Cells[c.Key] = c.Config
	}
	return state, nil
}

// Handle removes the row at the given index and drops every cell it
// referenced. No-op if the index is out of range.
func (p RemoveRowPayload) Handle(state Table) (Table, error) {
	idx := int(p.Index)
	if idx >= len(state.Rows) {
		return state, nil
	}
	for _, k := range state.Rows[idx].Cells {
		delete(state.Cells, k)
	}
	state.Rows = slices.Delete(state.Rows, idx, idx+1)
	return state, nil
}

// Handle inserts a column at the given index, mirroring AddRow with axes
// swapped. Cells whose row index exceeds the row count enter the cells map
// without being referenced by any row.
func (p AddColPayload) Handle(state Table) (Table, error) {
	cells := p.Cells
	if p.CellTemplate != nil {
		n := len(state.Rows)
		if n == 0 && len(state.Columns) == 0 {
			n = 1
		}
		cells = expandTemplate(*p.CellTemplate, n)
	}
	if len(state.Rows) == 0 && len(cells) > 0 {
		state.Rows = slices.Repeat([]Row{{Size: baseRowDim}}, len(cells))
	}
	idx := min(int(p.Index), len(state.Columns))
	state.Columns = slices.Insert(state.Columns, idx, Column{Size: max(p.Size, minCellDim)})
	if state.Cells == nil {
		state.Cells = make(map[string]CellConfig, len(cells))
	}
	for i := range state.Rows {
		if i >= len(cells) {
			break
		}
		rowIdx := min(idx, len(state.Rows[i].Cells))
		state.Rows[i].Cells = slices.Insert(state.Rows[i].Cells, rowIdx, cells[i].Key)
	}
	for _, c := range cells {
		state.Cells[c.Key] = c.Config
	}
	return state, nil
}

// Handle removes the column at the given index and drops every cell that
// column referenced across every row. No-op if the index is out of range.
func (p RemoveColPayload) Handle(state Table) (Table, error) {
	idx := int(p.Index)
	if idx >= len(state.Columns) {
		return state, nil
	}
	state.Columns = slices.Delete(state.Columns, idx, idx+1)
	for i := range state.Rows {
		if idx >= len(state.Rows[i].Cells) {
			continue
		}
		delete(state.Cells, state.Rows[i].Cells[idx])
		state.Rows[i].Cells = slices.Delete(state.Rows[i].Cells, idx, idx+1)
	}
	return state, nil
}

// Handle resizes the row at the given index. Sizes below the minimum cell
// dimension are clamped up to the floor. No-op if the index is out of range.
func (p ResizeRowPayload) Handle(state Table) (Table, error) {
	idx := int(p.Index)
	if idx >= len(state.Rows) {
		return state, nil
	}
	state.Rows[idx].Size = max(p.Size, minCellDim)
	return state, nil
}

// Handle resizes the column at the given index. Sizes below the minimum
// cell dimension are clamped up to the floor. No-op if the index is out of
// range.
func (p ResizeColPayload) Handle(state Table) (Table, error) {
	idx := int(p.Index)
	if idx >= len(state.Columns) {
		return state, nil
	}
	state.Columns[idx].Size = max(p.Size, minCellDim)
	return state, nil
}

// Handle replaces the cell config stored under p.Cell.Key. No-op if no entry
// with that key exists.
func (p SetCellPayload) Handle(state Table) (Table, error) {
	if _, ok := state.Cells[p.Cell.Key]; !ok {
		return state, nil
	}
	state.Cells[p.Cell.Key] = p.Cell.Config
	return state, nil
}

// Handle erases the selected cells. Fully-selected rows and columns are
// removed entirely; surviving cells in the selection have their variant and
// props replaced with the template's, keeping their original keys. Cells
// not in state.Cells are silently skipped.
func (p EraseCellsPayload) Handle(state Table) (Table, error) {
	if len(p.Cells) == 0 {
		return state, nil
	}
	selected := set.New(p.Cells...)
	type pos struct{ row, col int }
	posOf := make(map[string]pos)
	for r, row := range state.Rows {
		for c, k := range row.Cells {
			posOf[k] = pos{row: r, col: c}
		}
	}
	rowTally := make(map[int]int)
	colTally := make(map[int]int)
	for k := range selected {
		p, ok := posOf[k]
		if !ok {
			continue
		}
		rowTally[p.row]++
		colTally[p.col]++
	}
	fullRowIdx := []int{}
	for rowIdx, tally := range rowTally {
		if tally > 0 && tally == len(state.Rows[rowIdx].Cells) {
			fullRowIdx = append(fullRowIdx, rowIdx)
		}
	}
	slices.Sort(fullRowIdx)
	fullColIdx := []int{}
	for colIdx, tally := range colTally {
		if tally == len(state.Rows) {
			fullColIdx = append(fullColIdx, colIdx)
		}
	}
	slices.Sort(fullColIdx)
	for _, idx := range slices.Backward(fullRowIdx) {
		for _, k := range state.Rows[idx].Cells {
			delete(state.Cells, k)
		}
		state.Rows = slices.Delete(state.Rows, idx, idx+1)
	}
	for _, idx := range slices.Backward(fullColIdx) {
		state.Columns = slices.Delete(state.Columns, idx, idx+1)
		for r := range state.Rows {
			if idx >= len(state.Rows[r].Cells) {
				continue
			}
			delete(state.Cells, state.Rows[r].Cells[idx])
			state.Rows[r].Cells = slices.Delete(state.Rows[r].Cells, idx, idx+1)
		}
	}
	for _, k := range p.Cells {
		if _, ok := state.Cells[k]; !ok {
			continue
		}
		state.Cells[k] = p.Template
	}
	return state, nil
}
